src/recluster.c

```diff
@@ -0,0 +1,1175 @@
+		 * somebody is executing a database-wide CLUSTER), because there is
+		 * another check in cluster() which will stop any attempt to cluster
+		 * remote temp tables by name.  There is another check in
+		 * timescale_cluster_rel which is redundant, but we leave it for extra
+		 * safety.
+		 */
+		if (RELATION_IS_OTHER_TEMP(OldHeap))
+		{
+			relation_close(OldHeap, ExclusiveLock);
+			return;
+		}
+
+		if (OidIsValid(indexOid))
+		{
+			/*
+			 * Check that the index still exists
+			 */
+			if (!SearchSysCacheExists1(RELOID, ObjectIdGetDatum(indexOid)))
+			{
+				relation_close(OldHeap, ExclusiveLock);
+				return;
+			}
+
+			/*
+			 * Check that the index is still the one with indisclustered set.
+			 */
+			tuple = SearchSysCache1(INDEXRELID, ObjectIdGetDatum(indexOid));
+			if (!HeapTupleIsValid(tuple))	/* probably can't happen */
+			{
+				relation_close(OldHeap, ExclusiveLock);
+				return;
+			}
+			indexForm = (Form_pg_index) GETSTRUCT(tuple);
+			if (!indexForm->indisclustered)
+			{
+				ReleaseSysCache(tuple);
+				relation_close(OldHeap, ExclusiveLock);
+				return;
+			}
+			ReleaseSysCache(tuple);
+		}
+	}
+
+	/*
+	 * We allow VACUUM FULL, but not CLUSTER, on shared catalogs.  CLUSTER
+	 * would work in most respects, but the index would only get marked as
+	 * indisclustered in the current database, leading to unexpected behavior
+	 * if CLUSTER were later invoked in another database.
+	 */
+	if (OldHeap->rd_rel->relisshared)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot recluster a shared catalog")));
+
+	/*
+	 * Don't process temp tables of other backends ... their local buffer
+	 * manager is not going to cope.
+	 */
+	if (RELATION_IS_OTHER_TEMP(OldHeap))
+	{
+		if (OidIsValid(indexOid))
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot recluster temporary tables of other sessions")));
+		else
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot vacuum temporary tables of other sessions")));
+	}
+
+	/*
+	 * Also check for active uses of the relation in the current transaction,
+	 * including open scans and pending AFTER trigger events.
+	 */
+	/* TODO is this check needed/valid? */
+	CheckTableNotInUse(OldHeap, "CLUSTER");
+
+	/* Check heap and index are valid to cluster on */
+	check_index_is_clusterable(OldHeap, indexOid, recheck, ExclusiveLock);
+
+	/*
+	 * Quietly ignore the request if this is a materialized view which has not
+	 * been populated from its query. No harm is done because there is no data
+	 * to deal with, and we don't want to throw an error if this is part of a
+	 * multi-relation request -- for example, CLUSTER was run on the entire
+	 * database.
+	 */
+	if (OldHeap->rd_rel->relkind == RELKIND_MATVIEW &&
+		!RelationIsPopulated(OldHeap))
+	{
+		relation_close(OldHeap, ExclusiveLock);
+		return;
+	}
+
+	/*
+	 * All predicate locks on the tuples or pages are about to be made
```

Comment:
Can you look more into this. What exactly are predicate locks?

Reply:
(basing on [this](https://wiki.postgresql.org/wiki/Serializable#Predicate_Locking)) it's a fine grained lock (can lock individual tuples) which is used to guarantee serializable transactions don't read entries which were moved into its WHERE clause after the transaction's serialization point. I'm not touching this without reading a whole lot more of the postgres lock implementation.